### ISSUE TITLE
Automatic import (auto-detecting the import ID)

### ISSUE THREAD

**Issue by awwright** [hashicorp/terraform#27366]
### Current Terraform Version

```
Terraform v0.14.2
```

### Use-cases

There are cases where I have a lot of resources to import, and many of the resources only have a single valid ID I could possibly import with. I shouldn't have to compute this value in order to import this resource.

### Attempted Solutions

The only known work-around is to go to the documentation, determine the resource ID manually, then import it.

If there were a way to query the values that the resource would be created with, I could probably write an import script. However, I can't find any way to do a query like this, either.

### Proposal

There are two parts to this proposal:

1. If the resource would be created with a specific ID, then the `ID` argument should default to that.

2. An `--auto` flag (or similar name) should import _all_ resources that are not imported, and whose ID is unambiguous runtime. If the `ADDR` argument imported a resource, this will be performed afterwards.

### Example

For example, consider this Terraform document:

```terraform
resource "aws_s3_bucket" "example" {
   bucket = "my_bucket_example"
}

resource "aws_s3_bucket_public_access_block" "example" {
   bucket = aws_s3_bucket.example.id
   block_public_acls = true
   block_public_policy = true
}
```

Normally to import these resources, I would run:

```
terraform import aws_s3_bucket.example my_bucket_example
terraform import aws_s3_bucket_public_access_block.example my_bucket_example
```

Since we can determine what the ID must be, I should be able to omit this value, and get the same behavior:

```
terraform import aws_s3_bucket.example
terraform import aws_s3_bucket_public_access_block.example
```

Alternatively, I could run `terraform import --auto`, and the resource IDs will be detected because they are not in the state, and they have a deterministic ID.

Sometimes, the ID cannot be automatically computed. Consider if I use `random_string` to generate a bucket name:

```terraform
resource "random_string" "example" {
  length = 4
  special = false
  upper = false
}

resource "aws_s3_bucket" "example" {
   bucket = "my-bucket-${random_string.example.result}"
}

resource "aws_s3_bucket_public_access_block" "example" {
   bucket = aws_s3_bucket.example.id
   block_public_acls = true
   block_public_policy = true
}
```
Since none of the resources have a deterministic ID, `terraform import --auto` would not do anything by itself. However, once I import `random_string.example`, then the other resources now have a deterministic ID, and would be automatically imported:

```
$ terraform import random_string.example abcd --auto
random_string.example: Importing from ID "abcd"...
random_string.example: Import prepared!
random_string.example: Refreshing state [id=abcd]
aws_s3_bucket.example: Importing from ID "my-bucket-abcd"...
aws_s3_bucket.example: Import prepared!
aws_s3_bucket.example: Refreshing state [id=my-bucket-abcd]
aws_s3_bucket_public_access_block.example: Importing from ID "my-bucket-abcd"...
aws_s3_bucket_public_access_block.example: Import prepared!
aws_s3_bucket_public_access_block.example: Refreshing state [id=my-bucket-abcd]
```

### References

- #22219 — With this issue, the number of resources that would need to be specified in a bulk import would be reduced.

**Comment by jbardin** [comment]
Thanks @awwright 

Closing as a duplicate of #26364 so that we can more easily track the feature in a single issue. The concept also requires providers be able to indicate which attribute is a unique identifier in the first place, but modifying the protocol to handle that case would be part #26364 and #9048 anyway.

**Comment by awwright** [comment]
@jbardin I do not think that #26364 is a duplicate. If there are technical similarities, I'm unclear what those are, and I have a very different use-case that I would like to be considered.

> the concept also requires providers be able to indicate which attribute is a unique identifier in the first place

I was under the impression that you can generate an import ID, since Terraform is able to detect if that resource exists before it's added to the state.

I'm also not sure how this relates to #9048, unless you're proposing an alternate solution where Terraform automatically detects when resources get a new address (because their import ID remains the same).

**Comment by jbardin** [comment]
Terraform is not able to detect if a resource exists, only the provider can determine this. The providers also currently have no way to indicate which attributes are unique identifiers, so there would be no way for terraform to convert the configuration into an import request. Extending the API to allow providers to indicate unique identifiers would be how this relates to 9048. 

Since the goal for both this an 26364 is to allow import via the configuration, and would be resolved by the same solution, we can track them in the same issue. The issues are now linked, so any additional insight provided here can easily be view from the single open issue.

**Comment by kmoe** [comment]
Reopening as a useful statement of the "import ID problem".
